tests/test_builder.py

```diff
@@ -18,33 +27,128 @@
+            thread = get_thread(callback, parent)
+        self.assertEqual(thread.experimentPath, EXPERIMENT_PATH)
+        self.assertEqual(thread.experimentArgs, EXPERIMENT_ARGS)
+        self.assertEqual(thread.schedOpts, SCHED_OPTS)
+        mocked_submitted.connect.assert_called_once_with(callback, type=Qt.QueuedConnection)
 
-class _StringEntryTest(unittest.TestCase):
-    """Unit tests for _StringEntry class."""
+    def test_run(self):
+        self.mocked_response.json.return_value = 100
+        callback = mock.MagicMock()
+        parent = QObject()
+        with mock.patch("iquip.apps.builder.ExperimentSubmitThread.submitted") as mocked_submitted:
+            thread = get_thread(callback, parent)
+            thread.run()
+            thread.wait()
+        params = {
+            "file": EXPERIMENT_PATH,
+            "args": json.dumps(EXPERIMENT_ARGS),
+            **SCHED_OPTS
+        }
+        self.mocked_get.assert_called_once_with("http://127.0.0.1:8000/experiment/submit/",
+                                                 params=params,
+                                                 timeout=10)
+        mocked_submitted.emit.assert_called_once_with(100)
 
+    def test_run_request_exception(self):
+        """Tests when a requests.exceptions.RequestException occurs."""
+        self.mocked_response.raise_for_status.side_effect = requests.exceptions.RequestException()
+        callback = mock.MagicMock()
+        parent = QObject()
+        with mock.patch("iquip.apps.builder.ExperimentSubmitThread.submitted") as mocked_submitted:
+            thread = get_thread(callback, parent)
+            thread.run()
+            thread.wait()
+        params = {
+            "file": EXPERIMENT_PATH,
+            "args": json.dumps(EXPERIMENT_ARGS),
+            **SCHED_OPTS
+        }
+        self.mocked_get.assert_called_once_with("http://127.0.0.1:8000/experiment/submit/",
+                                                 params=params,
+                                                 timeout=10)
+        mocked_submitted.emit.assert_not_called()
 
-class _DateTimeEntryTest(unittest.TestCase):
-    """Unit tests for _DateTimeEntry class."""
+    def test_run_type_error(self):
+        """Tests when a TypeError occurs."""
+        callback = mock.MagicMock()
+        parent = QObject()
+        with mock.patch("iquip.apps.builder.ExperimentSubmitThread.submitted") as mocked_submitted:
+            experimentArgs = {"arg1": lambda: None}  # Not JSONifiable.
+            thread = get_thread(callback, parent, experimentArgs)
+            thread.run()
+            thread.wait()
+        self.mocked_get.assert_not_called()
+        mocked_submitted.emit.assert_not_called()
 
 
-class ExperimentSubmitThreadTest(unittest.TestCase):
-    """Unit tests for ExperimentSubmitThread class."""
+def get_thread(
+        callback: Callable[[int], None],
+        parent: QObject,
```

Comment:
Isn't it `Optional[QObject]`?

Reply:
Oh, I forgot removing this. I will fix it.

Reply:
Well, if you want to force to pass the `parent`, you can.
However, in this code, `parent` is only used as the `parent` parameter of `ExperimentSubmitThread`.
In its constructor, `parent` is `Optional[QObject]`.

Reply:
Then, I agree with you that setting `parent` as `Optional[QObject]` is more natural.
I will apply it.